src/module.c

```diff
@@ -12210,6 +12210,141 @@ int RM_LoadConfigs(RedisModuleCtx *ctx) {
+ * the RDB file will be loaded.
+ *
+ * `flags` must be zero. This parameter is for future use.
+ *
+ * On success REDISMODULE_OK is returned, otherwise
+ * REDISMODULE_ERR is returned and errno is set to the following values:
+ *
+ * * EINVAL: `stream` is NULL or `flags` value is invalid.
+ * * EIO: Failed to load the RDB file. Check server logs for more info.
+ *
+ * Example:
+ *
+ *     RedisModuleRdbStream *s = RedisModule_RdbStreamCreateFromFile("exp.rdb");
+ *     RedisModule_RdbLoad(s, 0);
+ *     RedisModule_RdbStreamFree(s);
+ */
+int RM_RdbLoad(RedisModuleRdbStream *stream, int flags) {
+    errno = 0;
+
+    if (!stream || flags != 0) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    }
+
+    if (server.aof_state != AOF_OFF) stopAppendOnly();
+
+    /* Kill existing RDB fork as it is saving outdated data. Also killing it
+     * will prevent COW memory issue. */
+    if (server.child_type == CHILD_TYPE_RDB) killRDBChild();
+
+    emptyData(-1,EMPTYDB_NO_FLAGS,NULL);
+
+    /* rdbLoad() can go back to the networking and process network events. If
+     * RM_RdbLoad() is called inside a command callback, we don't want to
+     * process the current client. Otherwise, we may free the client or try to
+     * process next message while we are already in the command callback. */
+    if (server.current_client) protectClient(server.current_client);
+
+    serverAssert(stream->type == REDISMODULE_RDB_STREAM_FILE);
+    int ret = rdbLoad(stream->data.filename,NULL,RDBFLAGS_NONE);
+
+    if (server.current_client) unprotectClient(server.current_client);
+    if (server.aof_state != AOF_OFF) startAppendOnly();
+
+    if (ret != RDB_OK) {
+        errno = EIO;
+        return REDISMODULE_ERR;
+    }
+
+    errno = 0;
+    return REDISMODULE_OK;
+}
+
+/* Save dataset to the RDB stream.
+ *
+ * `flags` must be zero. This parameter is for future use.
+ *
+ * If there is already a fork saving to the RDB file e.g. bgsave is in progress,
+ * then, module should guarantee that `stream` does not point to the same file.
+ * Otherwise, this function and RDB fork will try to write to the same file.
+ *
+ * On success REDISMODULE_OK is returned, otherwise
+ * REDISMODULE_ERR is returned and errno is set to the following values:
+ *
+ * * EINVAL: `stream` is NULL or `flags` value is invalid.
+ * * EIO: Failed to load the RDB file. Check server logs for more info.
+ *
+ * Example:
+ *
+ *     RedisModuleRdbStream *s = RedisModule_RdbStreamCreateFromFile("exp.rdb");
+ *     RedisModule_RdbSave(s, 0);
+ *     RedisModule_RdbStreamFree(s);
+ */
+int RM_RdbSave(RedisModuleRdbStream *stream, int flags) {
+    errno = 0;
+
+    if (!stream || flags != 0) {
+        errno = EINVAL;
+        return REDISMODULE_ERR;
+    }
+
+    serverAssert(stream->type == REDISMODULE_RDB_STREAM_FILE);
+
+    rdbSaveInfo rsi, *rsiptr;
+    rsiptr = rdbPopulateSaveInfo(&rsi);
+
+    if (rdbSave(SLAVE_REQ_NONE,stream->data.filename,rsiptr) != C_OK) {
```

Comment:
https://github.com/redis/redis/blob/ad72cb77975a54bc1c4b1370835c7803b1ce9439/src/rdb.c#L1502

```c
    server.dirty = 0;
    server.lastsave = time(NULL);
    server.lastbgsave_status = C_OK;
    stopSaving(1);
```
I have a question here. I call `rdbSave()` here directly and it sets above variables internally. 
If we don't see this new `RM_RdbSave()` as part of Redis persistence mechanism, probably we should not set those above variables? Should we even call callbacks through startSaving()/stopSaving()?

@oranagra wdyt?

Reply:
you're right. since we don't save the rdb file to the designated persistence location, then we should not mess up the status metrics (dirty, lastsave, etc).
but we should still call startSaving, to set the current state.
i.e. modules should be aware that we're going into a state of generating persistence (they may need to do some prep before being called to serialize their data)

Reply:
so i see you moved startSaving and stopSaving to the outer function, so that concern is resolved (no implications on the existing behavior, right?)
i would be ok with putting them inside the inner function too, but no objections to what you've done